app/lib/tool/mirror/mirror_config.dart

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+
+import 'package:json_annotation/json_annotation.dart';
+import 'package:meta/meta.dart';
+import 'package:yaml/yaml.dart' as yaml;
+
+part 'mirror_config.g.dart';
+
+/// The configuration to use when creating a local (partial) mirror of pub.dev
+@JsonSerializable(explicitToJson: true, includeIfNull: false)
+class MirrorConfig {
+  List<Package> packages;
+  List<Publisher> publishers;
+  List<User> users;
+  String defaultUser;
+
+  MirrorConfig({
+    @required this.packages,
+    @required this.publishers,
+    @required this.users,
+    this.defaultUser,
+  });
+
+  factory MirrorConfig.fromJson(Map<String, dynamic> json,
+      {bool normalize = false}) {
+    final mc = _$MirrorConfigFromJson(json);
+    if (normalize) mc.normalize();
+    return mc;
+  }
+
+  factory MirrorConfig.fromYaml(String source, {bool normalize = false}) {
+    final map = json.decode(json.encode(yaml.loadYaml(source)));
+    final mc = MirrorConfig.fromJson(map as Map<String, dynamic>);
+    if (normalize) mc.normalize();
+    return mc;
+  }
+
+  Map<String, dynamic> toJson() => _$MirrorConfigToJson(this);
+
+  void normalize() {
+    packages ??= <Package>[];
+    publishers ??= <Publisher>[];
+    users ??= <User>[];
+
+    for (final publisher in publishers) {
+      publisher.members ??= <Member>[];
+
+      // copy package definitions to [MirrorConfig.packages]
+      if (publisher.packages != null) {
+        publisher.packages.forEach((p) {
+          p.publisher = publisher.name;
+          packages.add(p);
+        });
+        publisher.packages = null;
+      }
```

Comment:
I propose we omit this... I see that with Dart typing we don't want to have two kinds of `Package` objects, and then this structure is not ideal.